### PR TITLE
Bug Fixes for PTQ and ACQ based OpenVINO Model Export and Added Test Cases

### LINKED ISSUE #2592: [Bug]:  ValueError while exporting model in OpenVINO format with PTQ and ACQ
### Describe the bug

Hello,

I have tried to export the pre-trained `PADiM` model in OpenVINO format with Post Training Quantization (PTQ) and Accuracy Control Quantization (ACQ).
During execution, I got a `ValueError` that says `'ImageBatch' object is not subscriptable`.

### Dataset

MVTecAD

### Model

PADiM

### Steps to reproduce the behavior

1. Install `anomalib` using `pip`.
2. Execute the following command,
```
AnomalibCLI(
    args=[
        "export",
        "--export_type",
        ExportType.OPENVINO,
        "--compression_type",
        CompressionType.INT8_PTQ,
        "--data",
        MVTec,
        "--data.root",
        f"{project_path}/Padim/MVTec
        "--data.category",
        "dummy",
        "--model",
        "Padim",
        "--default_root_dir",
        str(project_path),
        "--trainer.max_epochs",
        "1",
        "--ckpt_path",
        f"{project_path}/Padim/MVTec/dummy/v0/weights/lightning/model.ckpt",
    ],
)
```

### OS information

OS information:
- OS: [e.g. Ubuntu 20.04]
- Python version: [e.g. 3.10.0]
- Anomalib version: [e.g. 0.3.6]
- PyTorch version: [e.g. 1.9.0]
- CUDA/cuDNN version: [e.g. 11.1]
- GPU models and configuration: [e.g. 2x GeForce RTX 3090]
- Any other relevant information: [e.g. I'm using a custom dataset]

### Expected behavior

The model should be converted to OpenVINO format with PTQ which can be verified by successful execution.

### Screenshots

_No response_

### Pip/GitHub

pip

### What version/branch did you use?

main

### Configuration YAML

```yaml
No Configuration YAML since AnomaliB CLI have been used.
```

### Logs

```shell
The error arises from `models/components/base/export_mixin.py`
```

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by srikesh-07**
#### ✅ Fixes and Refactors
- **Fixed #2592** by **refactoring the access pattern** of `ImageBatch` from **subscriptable access** (e.g., `batch['image']`) to **attribute-style access** (e.g., `batch.image`).  
---
#### ✅ Addressing #2593: Metric Instantiation Issue in Model Export
- **#2593** was caused because the **`metric` parameter** was directly passed as a **string (`str`)** from the **CLI** into the `Model.export` function.
- This behavior required the following major modifications:

#### 1. Added `metric` as a subclass argument in `AnomalibCLI`
- Previously, the `metric` parameter was passed directly as a `str` in the CLI, causing failures when trying to instantiate it in `Model.export`.
- To solve this, the `metric` parameter is now added as a **subclass argument** in the `AnomalibCLI`.

#### 2. Migrated `metric` from `torchmetrics.Metric` to `anomalib.metrics.AnomalibMetric`
- This PR also aims to **migrate the metric's base class** from `torchmetrics.Metric` to `anomalib.metrics.AnomalibMetric` for seamless integration with the `Anomalib` ecosystem.

#### 3. Deferred Metric Instantiation in `AnomalibCLI`
- Since `AnomalibMetric` requires a mandatory `fields` parameter, metric instantiation has been **deferred from `AnomalibCLI`**.
- The `metric` is now instantiated within the `Model.export` function using the helper functions provided.

#### 4. Added Helper Functions for Metric Instantiation
- Introduced **helper functions** for `AnomalibMetric` which is similar to `AnomalibModel` that allow metric instantiation from `Model.export` via:
 
   - **String (`str`)**:
     ```shell
     --metric F1Score
     ```  
   - **Dictionary (`dict`)** input in Python:
     ```python
     {
         "classpath": "F1Score",
         "init_args": {
             "fields": ["pred_scores", "gt_label"]
         }
     }
     ```
   - **Namespace (`Namespace`)** passed in the pipeline via CLI.

#### 5. Allowed Dynamic Field Configuration via CLI
- Users can now pass **custom fields** via CLI to specify which fields should be used for metric calculation:
    ```shell
    --metric.fields "['pred_scores', 'gt_label']"
    ```

#### 6. Handled Missing Fields Gracefully
- In cases where users **do not pass `fields` explicitly**, the metric will use a placeholder:
    ```python
    fields = [""]
    ```
- The `fields` will later be modified depending on the **model type** (e.g., segmentation or classification).

#### 7. Modified ACQ Validation Function
-  The validation function utilized for ACQ is modified to support `AnomalibMetric`

---

####  ✅ Test Coverage Enhancements
- **Added unit test cases** for helper functions to instantiate the `AnomalibMetric`.
- **Added integration test cases** for OpenVINO export covering all compression types from:
    ```python
    from anomalib.export.deploy import CompressionType
    ```
- The compression types tested include:
    | Compression Type | Tested|
    |-----------------|---------|
    | **FP16**        | ✅ Yes    |
    | **INT8**        | ✅ Yes    |
    | **INT8_PTQ**    | ✅ Yes    |
    | **INT8_ACQ**    | ✅ Yes    |

---

#### ✅ PR Summary
- **Unblocks Model.export functionality** by allowing automatic metric instantiation in OpenVINO export.
- **Improves user experience** by allowing users to pass metric fields via CLI or Python without manual intervention.
- Fixed bugs related to `ImageBatch` access by switching from subscriptable access (`batch['image']`) to attribute-style access (`batch.image`).
- Ensures seamless compatibility with `AnomalibMetric`.

---

### 🛠️ Fixes
   - #2593 
   - #2592 
   - #2273

The CHANGELOG.md will be updated once the changes have been approved by the respective authors of the repository.

## ✨ Changes

Select what type of change your PR is:

- [X] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🔨 Refactor (non-breaking change which refactors the code base)
- [X] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation update
- [ ] 🔒 Security update

## ✅ Checklist

Before you submit your pull request, please make sure you have completed the following steps:

- [ ] 📋 I have summarized my changes in the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) and followed the guidelines for my type of change (skip for minor changes, documentation updates, and test enhancements).
- [X] 📚 I have made the necessary updates to the documentation (if applicable).
- [X] 🧪 I have written tests that support my changes and prove that my fix is effective or my feature works (if applicable).

For more information about code review checklists, see the [Code Review Checklist](https://github.com/openvinotoolkit/anomalib/blob/main/docs/source/markdown/guides/developer/code_review_checklist.md).

PS: This is my first PR to the Anomalib repository, and I sincerely hope that my contributions align with the standards and expectations set by the repository and its authors. I appreciate the time and effort of the authors who will be reviewing my PR, and I thank you in advance for considering my changes.

@samet-akcay @djdameln @ashwinvaidya17 

Thank You

### CODE REVIEW

**Comment by srikesh-07** [comment]
@samet-akcay @djdameln @ashwinvaidya17 

I see the build of ReadtheDocs has failed due to the **hash mismatch** during the pip installation. Can you please check out the issue ?

**Review by ashwinvaidya17 — COMMENTED** [review]
Thanks for the efforts. It is a really thorough PR and I appreciate detailed inline documentation. I am fine with the OpenVINO stuff but for the metrics part let's wait for @djdameln 's review as he is the expert here.

File: src/anomalib/models/components/base/export_mixin.py [hunk 1]
```diff
@@ -364,6 +364,11 @@ def _accuracy_control_quantization_ov(
             msg = "Metric must be provided for OpenVINO INT8_ACQ compression"
             raise ValueError(msg)
 
+        # Setting up the fields parameter in Metric if Metric is initialized with placeholder.
```
[the remark this answers is not in the document]
Comment (srikesh-07) [reply src/anomalib/models/components/base/export_mixin.py:367]: What I’m trying to convey is this:  

Let’s say the user passes a metric via the CLI using `--metrics f1_score` but does not specify any field parameters (i.e., `--metrics.fields` is not provided). In this case, `AnomalibMetric` is instantiated with a placeholder metric in the `model.export` method. Later, the metric’s fields are updated based on the model type in `export_mixin.py`.  

I also noticed a couple of grammatical mistakes in an inline comment and a logger statement:  

- In **[L367](https://github.com/openvinotoolkit/anomalib/blob/8a6ec1f76e7e009f3cf26aff11bbb2bc0f25ccb7/src/anomalib/models/components/base/export_mixin.py#L367)**:  
  Change **"if Metric is initialized with placeholder"** to **"Metric if Metric was initialized with a placeholder."**  

- In **[L370](https://github.com/openvinotoolkit/anomalib/blob/8a6ec1f76e7e009f3cf26aff11bbb2bc0f25ccb7/src/anomalib/models/components/base/export_mixin.py#L370)**:  
  Modify the logger statement from:  
  > `"The fields of metric are initialized empty. Setting it to model fields {metric.fields}"`  
  to:  
  > `"Since the fields of the metric were initialized as empty, they are now being set to the model fields: {metric.fields}."`  

I hope this clarifies your question. Let me know if you have any further doubts—I’d be happy to help!  

Thank you.

File: src/anomalib/metrics/__init__.py [hunk 2]
```diff
@@ -71,3 +79,184 @@
...
+        set[str]: Set of available metric names in snake_case format (e.g.
+            ``'min_max'``, ``'auroc'``, etc.)
+
+    Example:
+        Get all available metrics:
+
+        >>> from anomalib.metrics import get_available_metrics
+        >>> metrics = get_available_metrics()
+        >>> print(sorted(list(metrics)))  # doctest: +NORMALIZE_WHITESPACE
+            ['a_u_p_i_m_o', 'a_u_p_r', 'a_u_p_r_o', 'a_u_r_o_c',
```
[the remark this answers is not in the document]
Comment (srikesh-07) [reply src/anomalib/metrics/__init__.py:107]: Got it, I had the same thought. Can we assume that for single-word metrics, we capitalize the entire word, and for those with underscores, we use PascalCase? For example:  

- `"aupro"` → `"AUPRO"`  
- `"auroc"` → `"AUROC"`  
- `"dice"` → `"DICE"` (DICE may be implemented in the future)  
- `"f1_score"` → `"F1Score"`  
- `"min_max"` → `"MinMax"`  

Alternatively, we could map each metric in a hashmap within the `__init__.py` file under `anomalib.metrics`.

Let me know your thoughts on these approaches. 

Thank you.

[the remark this answers is not in the document]
Comment (samet-akcay) [reply src/anomalib/metrics/__init__.py:107]: @srikesh-07, yes, that's a good approach

[the remark this answers is not in the document]
Comment (srikesh-07) [reply src/anomalib/metrics/__init__.py]: @samet-akcay I've implemented the changes as discussed. Please let me know if any further modifications are needed. 

Thank you for your time and feedback.

**Comment by rajeshgangireddy** [comment]
Hi @srikesh-07 
Thanks for the contribution. I will take a look at this PR soon.

**Comment by rajeshgangireddy** [comment]
Hi @srikesh-07 
Thanks for the contribution. 
As anomalib has gone through a few changes after the creation of the PR, I would like to make several changes. 

So I would create a branch based on your branch and add our changes on top of them. That way your contributions are still counted.

**Comment by srikesh-07** [comment]
Thank you, @rajeshgangireddy. Please proceed with your changes. Let me know if there's anything else needed from my side.

**Comment by rajeshgangireddy** [comment]
Hi 
I would like to close this PR.
#3155 fixes this with minimal changes to existing code.
